Add metadata specification (#77)
neuroinformatics-unit/SWC-Blueprint: docs/source/metadata.md

> :orphan:

Reviewer: I assume this is a placeholder, but this page should be made far more accessible on the website eventually.

# Metadata

Metadata is additional data that describes the project data itself. Metadata can
be high-level (e.g. a general overview of the study and its purpose) or
low-level (acquisition parameters for extracellular electrophysiology
setup, or microscope).

> A number of detailed metadata standards exist, including

Reviewer: ```suggestion
A number of detailed metadata standards exist, including those from
```

[BIDS](https://bids-specification.readthedocs.io/en/stable/introduction.html),

> [openMinds](https://github.com/openMetadataInitiative) and

Reviewer: ```suggestion
[openMINDS](https://github.com/openMetadataInitiative) and
```

> [Allen](https://github.com/AllenNeuralDynamics/aind-data-schema),

Reviewer: ```suggestion
the [Allen Institute for Neural Dynamics](https://github.com/AllenNeuralDynamics/aind-data-schema),
```

each differing in its structure, level of detail and the datatypes they cover.

> Here, we provide a simple metadata organisation scheme that you can use to

Reviewer: ```suggestion
Here, we provide a simple schema that you can use to
```

> get started with adding metadata to your project. You are free to add

Reviewer: ```suggestion
get started with adding metadata to your NeuroBlueprint project. You are free to add
```

> metadata fields if you wish, but at the end of this guide we recommend fields

Reviewer: Maybe add a link?

that can go in each section.

> Please get in touch if you would like additional keys added to the metadata fields.

Reviewer: link to how they should contact us?

## Metadata Organisation Description

> At each level of the project, a metadata file can be included that describes that level:

Reviewer: Just seen that this comes later, but I think some intro would be better higher up in case people haven't come across yaml files.

```yaml
└── my_project/
    ├── my_project_metadata.yml
    └── rawdata/
        ├── rawdata_metadata.yml
        ├── sub-001/
        │   ├── sub-001_metadata.yml
        │   └── ses-001/
        │       ├── ses-001_metadata.yml
        │       ├── behav/
        │       │   └── behav_metadata.yml
        │       └── ephys/
        │           └── ephys_metadata.yml
        ├── sub-002/
        │   ├── sub-002_metadata.yml
        │   └── ...
        └── ...
```

**``project_metadata.yml``**
- This file contains high-level information about the project, for example its overall purpose,
who is involved in the project. See [project metadata](project-metadata).

**``rawdata_metadata.yml``**
- This file contains information about the data collection, for example the species of animal used
in the project. It may also contain specific sections for datatypes, that apply to all subjects

> in the project. For example, if `ephys` data was collected at a sampling rate of 30kHz for each subject,

Reviewer: ```suggestion
in the project. For example, if `ephys` data was collected at a sampling rate of 30kHz for every subject,
```

it may contain an `ephys` section with a `samplingRate` field. See [rawdata metadata](rawdata-metadata).

> **``sub-<value>_metadata.yml``**

Reviewer: Somewhere it should be clarified what the `<value>` refers to

in the project.

In this case, we can place the metadata entries for lower levels as keys at a higher level.
For example, if your `ephys` sampling rate for all subjects was `30 kHz`, you could structure
your `rawdata_metadata.yml` file as:

```yaml
SomeKey: someValue
ephys:
    samplingRate: 30000
```

This would then apply to all subjects in the `rawdata` folder.

However, this can be overwritten for particular cases e.g. if due to an error, a different
sampling rate was used in the acquisition. To do this, a metadata file should be included
for the case of interest in the relevant folder. For example, if `sub-005` used a different
sampling rate for all sessions, a `sub-005_metadata.yml` file could be included to overwrite the
information for this particular subject. e.g.

```yaml
samplingRate: 30500
notes: "A mistake was made during acquisition, leading to a sampling rate of 30500 Hz."
```

The folder structure may look like:

```
.
└── my_project/
    └── rawdata/
        ├── rawdata_metadata.yml   # contains the `ephys` entry applying to all subjects
        └── sub-001/
            ├── ses-001/
            │   └── ephys/

>             │       ├── ephys_metadata.yml   # contains the overwriting entry

Reviewer: overwriting seems like a weird word to use, but I don't have a better suggestion!

Author: 🤔

            │       └── ...
            └── ses-002/
                └── ...
```

> This was inspired by the similar inheritance principle in [BIDS](https://bids-validator.readthedocs.io/en/stable/validation-model/inheritance-principle.html)

Reviewer: ```suggestion
This was inspired by the similar inheritance principle in [BIDS](https://bids-validator.readthedocs.io/en/stable/validation-model/inheritance-principle.html).
```

# Recommended Metadata Keys

To ensure alignment across and within projects, we recommend using metadata keys from
a predefined set. Here we use BIDS as an existing source of metadata keys for each section.

> Please get in touch if you would like us to add new metadata fields to this list.

Reviewer: This is repeated, but useful. Perhaps it could be phrased in a different way?

(project-metadata)=
## Project Metadata

We use the BIDS `dataset_description.json` fields as a starting point for project-level metadata.

See the full specification for detailed descriptions:

> [BIDS Dataset Description Specification](https://bids-specification.readthedocs.io/en/stable/modality-agnostic-files/dataset-description.html#dataset_descriptionjson)

Reviewer: Missing full stop here and in the other, similar sentences.

> Recommended keys:

Reviewer: We should specify things like units here. There's no point standardising the names if some people use Hz and some kHz. I assume this is in the BIDS spec, but it would be good to make it clearer here.

Author: yes lets discuss this further at the meeting, I wasn't sure how best to do this. We can keep it very minimal here, or include quite a lot of the information from the BIDS site, or somewhere in between.

:orphan:
# Metadata

[Scientific metadata](https://www.dcc.ac.uk/resources/curation-reference-manual/chapters-production/scientific-metadata)
is additional information included with a project that describes the data itself. Adding metadata to your project
makes it easier to understand, reproduce and share the data.

Metadata can be high-level (e.g. a general overview of the study and its purpose) or
low-level (acquisition parameters for extracellular electrophysiology
setup, or microscope).

A number of detailed metadata standards exist, including those from
[BIDS](https://bids-specification.readthedocs.io/en/stable/introduction.html),
[openMINDS](https://github.com/openMetadataInitiative) and
the [Allen Institute for Neural Dynamics](https://github.com/AllenNeuralDynamics/aind-data-schema),
each differing in its structure, level of detail and the datatypes they cover.

Here, we provide a simple schema that you can use to

> get started with adding metadata to your NeuroBlueprint project. You are free to add

Reviewer 2: Elsewhere (e.g. in the folder structure specification page), we use bolded **NeuroBlueprint**, we should do the same in this page, for consistency.

metadata fields if you wish, but at the
[end of this guide](metadata-keys) we recommend fields
that can go in each section.

Please get in touch by raising a
[GitHub Issue](https://github.com/neuroinformatics-unit/NeuroBlueprint/issues)
if you would like additional keys added to the metadata fields.

# YAML file format

Metadata files should use the [YAML](https://yaml.org/) file format (`.yaml` extension).
YAML is a human-readable text format designed to be easy to read and edit.

YAML stores information as **key-value pairs** and uses indentation (spaces) to represent structure.

For example, a simple metadata file might look like:

```yaml
project:
    Name: "Visual Decision Making Study"
    Authors:
      - "Jane Smith"
      - "John Doe"

sub:
    Species: "Mus musculus"
    Genotype: "Thy1-GCaMP6s/wt"

ephys:
    SamplingFrequency: 30000
    ManufacturersModelName: "Neuropixels 2.0"
```

# Metadata organisation schema

Metadata should be stored in `metadata.yaml` files that can include any of

> these pre-defined sections, which map to NeuroBlueprint folder levels:

Reviewer 2: I think we should already make it clear at this point that there are two types of metadata-containing yaml files:

1. The `metadata.yaml` files mapping to one of the predefined NeuroBlueprint folder levels (project, sub, ses, or one of the datatypes). These apply to anything that's on the same level or below.
2. The `{filename}_metadata.yaml` 'sidecar' files that can be associate with a specific element (file or folder) stored *within* datatype folders. The apply *only* to the element whose name they are prefixed with.

What follows can stay largely the same, e.g. first explaining the first type of metadata files as you do, introducing the inheritance principle, and then explaining the second type as a way towards more fine-grained control (which still obeys inheritance).

But I think it could be useful to introduce that distinction upfront.

Author: I think this is good practice for giving a proper overview of the full spec. My only reservation is hopefully these `metadata.yml` files will cover 90% of use cases and it would be nice to funnel people into using them as much as possible. The sidecar metadata only really exist to cover edge cases and it's not something we have properly fleshed out, and it might be confusing to introduce both at the same time up front.

As such a super-simple entry point that funnels people into the workflow we'd rather they use might nice, and then power users who find the basic spec does not apply can extend using the sidecar section?

Author: Have included this in the section, funneling down from less specific (project, sub, datatype) or more specific (single folder)

    StrainRrid: "RRID:IMSR_JAX:000664"

ses:
    Experimenter: "Jane Smith"
    Notes: "First recording session"

ephys:
    SamplingFrequency: 30000
    ManufacturersModelName: "Neuropixels 2.0"
    PowerLineFrequency: 50
    RecordingType: "continuous"

behav:
    TaskName: "Visual Decision Making"
    TaskDescription: "Mouse makes decisions based on visual stimuli"
```

For more details on what information should be put in each section,
see the [metadata keys](metadata-keys) section.

The location of the `metadata.yaml` file indicates what data the metadata refers to.

**A metadata file applies to data in all folder levels at or below it.**

In the above example, the entries apply to all data in the project. Therefore,
we can place this metadata file at the top-level of the project:

```
└── my_project/
    ├── metadata.yaml
    └── rawdata/
        ├── sub-001/
        │   └── ses-001/
        │       ├── behav/
        │       └── ephys/
        ├── sub-002/
        │   └── ...
        └── ...
```

## Inheritance

In some cases, the same metadata might not apply to all data within the project.
In this case, a `metadata.yaml` file can be placed at a lower level, overwriting

> the metadata fields at a higher level.

Reviewer 2: May be clearer this way? 'At a higher level' may falsely imply that metadata as applied to the higher levels is being overridden.

```suggestion
Sometimes, the same metadata might not apply to all data within the project.
In such cases, a `metadata.yaml` file can be placed at a lower level, overwriting
the metadata fields inherited from higher levels.
```

For example, let's say that `sub-001/ses-001` used a different sampling frequency,
due to an experimental error. Instead of 30000 Hz, 27000 Hz was used.

We can put a new `metadata.yaml` file in the `sub-001/ses-001` folder
with the `ephys` entry and updated sampling frequency:

```
└── my_project/
    ├── metadata.yaml
    └── rawdata/
        ├── sub-001/
        │   └── ses-001/
        │       ├── metadata.yaml
        │       ├── behav/
        │       └── ephys/
        ├── sub-002/
        │   └── ...
        └── ...
```

with `sub-001/ses-001/metadata.yaml` containing:

```yaml
ephys:
  SamplingFrequency: 27000
```

Now, `SamplingFrequency` at 30000 applies to all sessions except for
`sub-001/ses-001`, which is 27000:

This was inspired by the similar inheritance principle in [BIDS](https://bids-validator.readthedocs.io/en/stable/validation-model/inheritance-principle.html)

:::{tip}
When it is equivalent to put the `metadata.yaml` file at one of multiple folder levels,
we recommend placing the file at the highest possible level.

> For example, above the ephys information for `sub-001/ses-001/metadata.yaml` is placed in the

Reviewer 2: Was momentarily confused by the 'above the ephys' phrase.

```suggestion
In the above example, the ephys information for `sub-001/ses-001/metadata.yaml` is placed in the
```

session folder, rather than the equally valid ephys folder `sub-001/ses-001/ephys/metadata.yaml`.
:::

# Sidecar metadata files

It may be required to associate metadata with a specific data file or folder.
In this case, a metadata YAML file that copies the original
file or folder name with the suffix `_metadata` can be used. These are called 'sidecar' metadata files.

> Sidecar metadata files should contain use the same sections as above (e.g. `sub`, `behav` etc.).

Reviewer 2: ```suggestion
Sidecar metadata files should contain the same sections as above (e.g. `sub`, `behav` etc.).
```

For example, in the `behav` folder there may be multiple
acquisition runs, with different associated metadata. In this case, a metadata
YAML file that applies specifically to a particular file can be created.

> For example, in the project below all videos are assumed to be of the task

Reviewer 2: We don't need to repeat 'For example' this much, the previous paragraph already opens with 'For example'.

```suggestion
In the project below all videos are assumed to be of the task
```

`"Visual Decision Making"`. However, imagine for the run `run-002_camera-top.mp4`
a video was taken of a different task.

In this case, we can use inheritance to overwrite the `TaskName` for this particular run
by placing a sidecar metadata file next to the video file of interest:

```
└── my_project/
    ├── metadata.yaml
    └── rawdata/
        └── sub-001/
            └── ses-001/
                └── behav/
                    ├── run-001_camera-top.mp4
                    ├── run-002_camera-top.mp4
                    └── run-002_camera-top_metadata.yaml
```

`/my_project/metadata.yaml`

```yaml
behav:
    TaskName: "Visual Decision Making"
```

`run-002_camera-top_metadata.yaml`

```yaml
behav:
    TaskName: "Visual Decision Making Modified Long"
```

:::{warning}

All metadata files must be called `metadata.yaml` or be
the exact name of an existing file or folder with a `_metadata` suffix.

For example, `my_recording_metadata.yaml` is only valid if there is a file or folder
called `my_recording` (e.g. `my_recording.bin`).

:::

(metadata-keys)=
# Suggested metadata keys

> NeuroBlueprint does not mandate the use of any specific metadata key-value

Reviewer 2: ...other than the highest-level 'section' keys (project, sub, ses, or one of the datatypes) mentioned above. Good to remind users of that.

Basically, you can use any keys you like, but they have to fall within one of these recognised sections.

pairs. In theory, the structural rules above can be used to document the project
as you wish.

> However, we highly recommend using one of the suggested metadata

Reviewer 2: Maybe 'a subset of the suggested' instead on 'one of the'? Typically it will be more than one, and 'subset' nicely communicates that they don't have to use the full suggested set.

Author: Thanks @niksirbi  that's a good point, I added subset but then wasn't sure if it looked like they strictly use a subset of the keys rather than all). I tried to make this explicit below, but maybe if its too work we can just use a 'subset' (or say 'a subset or all of the'

However, we highly recommend using the suggested metadata
keys detailed below if they cover the concept you want to document. 
This helps ensure that metadata entries are 
interoperable with your immediate colleagues, and across labs.
There is no requirement to use all keys, selecting a subset that work best for
your requirements is encouraged.

Reviewer 2: I think it's clear enought that way.

keys detailed below if it covers the concept you want to document. This helps
ensure that metadata entries are interoperable with your immediate colleagues, and across labs.

For example, if you want to include the species of mice in your project, it
makes sense to use the suggested key `Species` rather than come up with your own key.

If you find that your use case is not covered by a suggested key, please
[get in touch](https://github.com/neuroinformatics-unit/NeuroBlueprint)

> and we will be happy to add it to this page.

Reviewer 2: I'm not sure we want people contacting us for any possible custom key they want to add. 
Maybe instead we should reframe this slightly as: Get in touch if you think something else is really useful and should be part of the 'suggested' keys?

Reviewer: Also, maybe it should specify how they get in touch? That's just a link to the repo. Maybe for new keys, Zulip is the best way to get in touch?

Reviewer 2: I think this point is still outstanding.

## `project` metadata

This file contains high-level information about the project, for example its overall purpose,

> who is involved in the project.

Reviewer 2: We should now talk about *sections* not *files*. Also applied to subsequent sections.

```suggestion
This section contains high-level information about the project, for example the people involved and the project's overall purpose.
```

For detailed descriptions of the suggested keys, see the
[BIDS Dataset Description Specification](https://bids-specification.readthedocs.io/en/stable/modality-agnostic-files/dataset-description.html#dataset_descriptionjson).

Suggested keys:

```yaml
Name:
NeuroBlueprintVersion:
HEDVersion:
DatasetLinks:
DatasetType:
License:
Authors:
Keywords:
Acknowledgements:
HowToAcknowledge:
Funding:
EthicsApprovals:
ReferencesAndLinks:
DatasetDOI:
SourceDatasets:
```

(sub-metadata)=
## `sub` metadata

This metadata file contains information about experimental subjects,
for example the date of birth, identifiers, strain or genotype.

Suggested key are taken from the
[BIDS Participants Specification](https://bids-specification.readthedocs.io/en/stable/modality-agnostic-files/data-summary-files.html).

> Note they are presented as snake_case in the specification, but we recommend PascalCase for consistency with other metadata files.

Reviewer 2: I usually see this referred to as CamelCase (instead of PascalCase) but that may just be me.

Reviewer: I also call it camelCase, but technically this is correct, as camelCase starts lowercase, but PascalCase starts capitalised.

Suggested keys:

```yaml
SubjectId:
Age:
Sex:
Handedness:
Species:
Strain:
StrainRrid:
Genotype:
DateOfBirth:
```

(ses-metadata)=
## `ses` metadata

This file contains information related to the particular experimental session. For example,
the date, additional notes on what happened in the session.

We use the BIDS session fields as a starting point for session-level metadata.

Suggested key are taken from the
[BIDS Sessions Specification](https://bids-specification.readthedocs.io/en/stable/modality-agnostic-files/data-summary-files.html#sessions-file).
Note they are presented as snake_case in the specification, but we recommend PascalCase for consistency with other metadata files.

Suggested keys:

```yaml
SessionId:
SessionDate:
Age:
Weight:
Notes:
Experimenter:
```

(datatype-keys)=

> ## Datatype keys

Reviewer 2: I'm slightly worried about the suggested datatype keys listed here per datatype. There is a risk of them getting out of sync with the upstream linked BIDS / BEP specs. The dataset description, subject and session keys should not change much (if at all) so we should keep the suggested sets for those.

But datatypes are more fragile (especially the BEPs ones not yet merged).

How about, for each datatype we begin by linking to the relevant BIDS/BEP pages, as you already do, but then in a snippet we provide just a few example fields for each datatype, making it clear that this is not the full set?

Author: hmm this is true, my hesitation with linking to the spec is that it's quite confusing in terms of what are valid metadata keys, as they have the PascalCase keys for their .json (which we copy) but the snake case keys for the .json (which we don't). Rather than explaining this distinction and telling them to only use PascalCase it seems more natural to provide a list of keys and tell them to find them in the spec for more information on those specific keys.

Maybe we just include keys from merged BEPs (with the exception of electrophysiology as its nearly merged) and assume the key names themselves will not change? If they add keys but we miss them its not the end of the world?

Author: Adam mentioned that this would make versioning NB difficult as the keys could change on the BIDS side but we would not have a record of what they were. To combat this issue though, we can encode the NB spec in a .yaml file and then test against the BIDS yaml spec to check if keys have come out of alignment.

Datatype metadata files contain acquisition-specific metadata for each modality.
NeuroBlueprint contains many possible [datatypes](https://neuroblueprint.neuroinformatics.dev/latest/specification.html#datatype)
e.g. `ephys`, `behav`, `funcimg`, `anat`.

> Where possible, we recommend using the BIDS keys from relevant Bids Extension Protocols (BEPs).

Reviewer 2: ```suggestion
Where possible, we recommend using the BIDS keys from relevant Bids Extension Proposals (BEPs).
```

for detailed descriptions.

Suggested keys:

- **Institution Information**
  - `InstitutionName`
  - `InstitutionAddress`
  - `InstitutionalDepartmentName`

- **Setup Information**
  - `PowerLineFrequency`
  - `Manufacturer`
  - `ManufacturersModelName`
  - `ManufacturersModelVersion`
  - `RecordingSetupName`
  - `SamplingFrequency`
  - `DeviceSerialNumber`
  - `SoftwareName`
  - `SoftwareVersions`
  - `RecordingDuration`
  - `RecordingType`
  - `EpochLength`

- **Processing Information**
  - `SoftwareFilters`
  - `HardwareFilters`

- **Pharmaceuticals**
  - `PharmaceuticalName`
  - `PharmaceuticalDoseAmount`
  - `PharmaceuticalDoseUnits`
  - `PharmaceuticalDoseRegimen`
  - `PharmaceuticalDoseTime`

- **Sample**
  - `BodyPart`
  - `BodyPartDetails`
  - `BodyPartDetailsOntology`
  - `SampleEnvironment`
  - `SampleEmbedding`
  - `SliceThickness`
  - `SampleExtractionProtocol`

- **Supplementary**
  - `SupplementarySignals`

- **Task Information**
  - `TaskName`
  - `TaskDescription`
  - `Instructions`
  - `CogAtlasID`
  - `CogPOID`

- **Coordinate System**
  - `MicroephysCoordinateUnits`

(behav-metadata)=
### `behav`

See the

> [BIDS Behavioural Experiments Specification](https://bids-specification.readthedocs.io/en/stable/modality-specific-files/behavioral-experiments.html).

Reviewer 2: There are actually two relecant BIDS specs here. Apart from the established `beh` datatype you link to, there is a new proposal underway for Behavioural audio/video recordings, which may be way more relevant for systems neuro users.

See https://bids.neuroimaging.io/extensions/beps/bep_047.html

Author: we could skip this for now [re: here](https://github.com/neuroinformatics-unit/NeuroBlueprint/pull/77#discussion_r3357139996)?

Author: Let's add as an issue

for detailed descriptions.

Suggested keys:

**Institution Information**
- `InstitutionName`
- `InstitutionAddress`
- `InstitutionalDepartmentName`

> -

Reviewer 2: Orphaned bullet point
```suggestion

```

    Authors:
      - "Jane Smith"
      - "John Doe"

sub:
    Species: "mus musculus"
    Genotype: "Thy1-GCaMP6s/wt"

ephys:
    SamplingFrequency: 30000
    ManufacturersModelName: "Neuropixels 2.0"
```

## Metadata organisation schema

Metadata should be stored in `metadata.yaml` files that can include any of
these pre-defined sections, which map to [**NeuroBlueprint** folder levels](neuroblueprint-folders):

- `project`
- `sub`
- `ses`
- [**NeuroBlueprint** datatype](https://neuroblueprint.neuroinformatics.dev/latest/specification.html#datatype) (e.g. `behav`, `ephys`)

`metadata.yaml` files may be placed at any of the corresponding folder levels and
can contain information for that folder level and below. To include metadata on
a particular file, a [sidecar metadata file](sidecar-files) can be placed next to the file.

For example, the contents of a `metadata.yaml` placed in the project folder root might look like:

```yaml
project:
    Name: "Visual Decision Making Study"
    Authors:
      - "Jane Smith"
      - "John Doe"

sub:
    Species: "mus musculus"
    Strain: "C57BL/6J"
    StrainRrid: "RRID:IMSR_JAX:000664"

ses:
    Experimenter: "Jane Smith"

>     Notes: "First recording session"

Reviewer: Does this make sense? Inheritance would mean that every session is "first recording session". Is there something else that could go in the notes?

Author: yes that is not a good example, I added `"Visual decision-making recording sessions using protocol 3A"`

Its a bit tricky, it needs to be detailed enough for `ses`, not too detailed such that it should go in a datatype folder, but apply to all sessions. I think that works if the reference to the mysterious Protocol 3A is not confusing

Reviewer: Yeah I think your new version is better. We can iterate on these examples if we think of something even better later.

Suggested keys with example entries:

```yaml
project:
    Name: "Cortical Dynamics of Visual Decision Making in Mice"
    Authors:
      - "Jane Smith"
      - "John Doe"
    Keywords:
      - "mouse"
      - "mus musculus"
      - "visual cortex"
      - "decision making"
      - "Neuropixels"
      - "extracellular electrophysiology"
      - "two-photon imaging"
    Abstract: >
      Head-fixed mice performed a two-alternative forced-choice visual
      discrimination task while neural activity was recorded from visual
      and parietal cortex using Neuropixels 2.0 probes. The dataset
      accompanies a study of how choice-related signals emerge across
      cortical areas during perceptual decision making.
    NeuroBlueprintVersion: "0.5.0"
    License: "CC-BY-4.0"
    Acknowledgements: >
      We thank the institute's animal facility staff for husbandry support.
    HowToAcknowledge: >
      Please cite Smith et al. (2026), Journal of Neuroscience,
      https://doi.org/10.1234/jneurosci.2026.0001
    Funding:
      - "Wellcome Trust Grant 219627/Z/19/Z"

>       - "BBSRC Grant BB/X001234/1"

Reviewer: Very picky, but is it worth avoiding listing grants we're not associated with? Just in case someone thinks BBSRC funded this?

Author: yes can't hurt to avoid confusion!
